File: frontend/src/api/subjectContent.ts
import { ApiError } from "./editions";
import type { PublicationDocument } from "./production";

export interface SubjectContentResponse {
  subject_id: string;
  run_id: string;
  pipeline_generation: number;
  artifact_id: string;
  artifact_version: number;
  artifact_input_hash: string;
  status: string;
  schema_version: string;
  canonical_content: PublicationDocument;
  rendered_content: string | null;
}

export interface SubjectIndicatorResponse {
  id: string;
  artifact_type: string;
  display_value: string;
  normalized_value: string;
  indicator_status: string;
  source_ids: string[];
}

export interface SubjectAssetResponse {
  id: string;
  original_name: string;
  mime_type: string | null;
  sha256: string | null;
  size: number | null;
  origin: string;
  provenance: Record<string, string> | null;
  tlp: string;
  do_not_submit: boolean;
  external_llm_allowed: boolean;
}

export interface SubjectAssetsResponse {
  sources: SubjectAssetResponse[];
  samples: SubjectAssetResponse[];
}

export function getSubjectContent(
  subjectId: string,
): Promise<SubjectContentResponse | null> {
  return requestOrNull(
    `/api/subjects/${encodeURIComponent(subjectId)}/content`,
  );
}

export function getSubjectIndicators(
  subjectId: string,
): Promise<SubjectIndicatorResponse[]> {
  return request(`/api/subjects/${encodeURIComponent(subjectId)}/indicators`);
}

export function getSubjectAssets(
  subjectId: string,
): Promise<SubjectAssetsResponse | null> {
  return requestOrNull(`/api/subjects/${encodeURIComponent(subjectId)}/assets`);
}

async function request<T>(url: string): Promise<T> {
  const response = await fetch(url);
  if (response.ok) return (await response.json()) as T;
  throw await apiError(response);
}

async function requestOrNull<T>(url: string): Promise<T | null> {
  const response = await fetch(url);
  if (response.status === 404) return null;
  if (response.ok) return (await response.json()) as T;
  throw await apiError(response);
}

async function apiError(response: Response): Promise<ApiError> {
  const body = (await response.json().catch(() => null)) as {
    detail?: { code?: string; message?: string } | string;
  } | null;
  const detail = body?.detail;
  const message =
    typeof detail === "string"
      ? detail
      : detail?.message || "Le contenu du sujet est inaccessible.";
  return new ApiError(
    message,
    typeof detail === "object" && detail?.code
      ? detail.code
      : "subject_content_error",
    response.status,
  );
}

File: backend/src/cti_app/api/subject_content.py
"""Stable Subject read endpoints."""

from __future__ import annotations

from typing import Any, cast
from uuid import UUID

from fastapi import APIRouter, HTTPException, Request, status
from pydantic import BaseModel, ConfigDict

from cti_app.application.production_parsers import IndicatorStatus
from cti_app.application.subject_content import (
    SubjectAssetsView,
    SubjectContentService,
    SubjectContentView,
    SubjectIndicatorView,
)
from cti_app.domain.classification import TLP
from cti_app.domain.production import ProductionArtifactStatus
from cti_app.domain.publication import ArtifactType

router = APIRouter(prefix="/api", tags=["subject-content"])


class ContentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    subject_id: UUID
    run_id: UUID
    pipeline_generation: int
    artifact_id: UUID
    artifact_version: int
    artifact_input_hash: str
    status: ProductionArtifactStatus
    schema_version: str
    canonical_content: dict[str, Any]
    rendered_content: str | None


class IndicatorResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    artifact_type: ArtifactType
    display_value: str
    normalized_value: str
    indicator_status: IndicatorStatus
    source_ids: tuple[str, ...]


class AssetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    original_name: str
    mime_type: str | None
    sha256: str | None
    size: int | None
    origin: str
    provenance: dict[str, str] | None
    tlp: TLP
    do_not_submit: bool
    external_llm_allowed: bool


class AssetsResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    sources: list[AssetResponse]
    samples: list[AssetResponse]


def _service(request: Request) -> SubjectContentService:
    return cast(SubjectContentService, request.app.state.subject_content_service)


def _not_found() -> HTTPException:
    return HTTPException(
        status_code=status.HTTP_404_NOT_FOUND,
        detail={
            "code": "subject_content_not_found",
            "message": "Aucun contenu produit n'est disponible pour ce sujet.",
        },
    )


@router.get("/subjects/{subject_id}/content", response_model=ContentResponse)
async def get_subject_content(subject_id: UUID, request: Request) -> ContentResponse:
    value: SubjectContentView | None = await _service(request).content(subject_id)
    if value is None:
        raise _not_found()
    return ContentResponse.model_validate(value, from_attributes=True)


@router.get("/subjects/{subject_id}/indicators", response_model=list[IndicatorResponse])
async def get_subject_indicators(
    subject_id: UUID, request: Request
) -> list[IndicatorResponse]:
    values: list[SubjectIndicatorView] = await _service(request).indicators(subject_id)
    return [IndicatorResponse.model_validate(value, from_attributes=True) for value in values]


@router.get("/subjects/{subject_id}/assets", response_model=AssetsResponse)
async def get_subject_assets(subject_id: UUID, request: Request) -> AssetsResponse:
    value: SubjectAssetsView | None = await _service(request).assets(subject_id)
    if value is None:
        raise _not_found()
    return AssetsResponse(
        sources=[
            AssetResponse.model_validate(item, from_attributes=True)
            for item in value.sources
        ],
        samples=[
            AssetResponse.model_validate(item, from_attributes=True)
            for item in value.samples
        ],
    )
